paprika/tleap.py

```diff
@@ -0,0 +1,574 @@
+                number_to_add = int(np.ceil(float(amount[:-1]) * self.target_waters * 0.018))
+                self.add_ion_residues.append(number_to_add)
+            elif isinstance(amount, str) and amount[-1] == 'M':
+                # User specifies molarity...
+                volume = self.get_volume()
+          ################# raise Exception if volume is None?????????? ########################################!!!!!!!!!!!!
+                number_of_atoms = float(amount[:-1]) * N_A
+                liters = volume * ANGSTROM_CUBED_TO_LITERS
+                number_to_add = int(np.ceil(number_of_atoms * liters))
+                self.add_ion_residues.append(number_to_add)
+            else:
+                raise Exception('Unanticipated error calculating how many ions to add.')
+
+    def get_volume(self):
+        """
+        Run and parse `tleap` output and return the volume of the structure.
+    
+        Returns
+        -------
+        volume : float
+            The volume of the structure in cubic angstroms
+
+        """
+        output = self.run()
+        # Return the total simulation volume
+        for line in output[0].splitlines():
+            line = line.strip()
+            if "Volume" in line:
+                match = re.search("Volume(.*)", line)
+                volume = float(match.group(1)[1:-4])
+                return volume
+        log.warning('Could not determine total simulation volume.')
+        return None
+
+    def remove_waters_manually(self):
+        """
+        Remove a few water molecules manually with `tleap` to exactly match a desired number of waters.
+
+        """
+    
+        cycle = 0
+        max_cycles = 10
+        waters = self.wat_added_history[-1]
+        while waters > self.target_waters:
+            # Retrieve excess water residues
+            water_surplus = (waters - self.target_waters)
+            water_residues = self.list_waters()
+            self.waters_to_remove = water_residues[-1 * water_surplus:]
+            log.debug('Manually removing waters... {}'.format(self.waters_to_remove))
+
+            # Get counts for all residues
+            residues = self.count_residues()
+
+            # Check if we reached target
+            waters = residues['WAT']
+            if waters == self.target_waters:
+                for key, value in sorted(residues.items()):
+                    log.info('{}\t{}'.format(key, value))
+                return
+            cycle += 1
+            if cycle > max_cycles:
+                raise Exception(
+                    "Solvation failed due to an unanticipated problem with water removal."
+                )
+
+    def list_waters(self):
+        """
+        Run and parse `tleap` output and return the a list of water residues.
+    
+        Returns
+        -------
+        water_residues : list
+            A list of the water residues in the structure
+
+        """
+        output = self.run()
```

Comment:
I wonder if we're running `tleap` and excess number of times. For example, if I just ran it once to solvate and I want the list of waters, why does it have to run `tleap` again?

Reply:
This was in the back of my mind too, but I left the current logic to make sure I passed the tests.  Before I finish pull request, I'll check to make sure we don't double run `run()`

Reply:
I think we're good, but maybe I missed something.